### DISCUSSION TITLE
Returning a value from a store function

### DISCUSSION THREAD

**Discussion by feederaccount** [pmndrs/zustand#1785] [Q&A] [answered]
I am using Immer with Zustand and I want to get the return type from a method when it is called on the store:

```
export const useStore = create<MyStoreState>()(devtools((set, get) => ({
    userList: new UserList(),
    addUser: (user: User): boolean => {
        set(produce((draft: MyStoreState) => {
            const addResult = draft.userList.addUser(user);
            // how to return addResult?
        }));
    },
})));
```

Is there a technique for doing this? I know having classes isn't all that common with Zustand, and maybe this is an OOP bias, but it's useful being able to call class methods like this and it works well with Immer.

The only solution I can see is storing the result on the store itself, so the store would have
```
userList: ...,
addUser: ...,
addUserResult: ...
```

I can't be the first person to do this.. are there any better approaches?

**Comment by dai-shi** [comment, marked as answer]
`set` run the function in sync. So, you can store the result in a local variable. But, I'm not sure if it's a good programming pattern with immer.

**Reply by feederaccount** [reply]
Yeah I decided against it in the end. I've gone for a different approach entirely as returning from this doesn't work well.

> I'm not sure if it's a good programming pattern with immer.

Do you mean my use of classes or returning from set?

**Reply by dai-shi** [reply]
I mean returning a value from immer `produce`.

**Comment by MathurAditya724** [comment]
Hey, I am trying to create a `uniqueId` generator in Zustand, here is part of the code - 

```ts
_unique: {},
uniqueId: (type, key = "default") => {
  let id = 1;

  // Generating new id
  set((state) => {
    // Getting cache or init a new one
    const context = state._unique[key] ?? {};

    // Starting from 1
    id = (context[type] ?? 0) + 1;

    // Updating the cache
    context[type] = id;
    state._unique[key] = context;
  });

  // Returning the new value
  return `${type}${id}`;
},
``` 

I am updating the cache and returning the new ID. I am using Immer also. But this is not working. Any Idea?

You can check out the repo at [https://github.com/rhinobase/fibr](https://github.com/rhinobase/fibr)

**Reply by dai-shi** [reply]
I think it should technically work, so I wonder if immer does something weird.

**Reply by MathurAditya724** [reply]
Thanks for the reply, I got it working. The problem came from `immer` and also my code was flawed, basically what I was trying to achieve was to have a local store to cache all the types of the app and their occurrence. So that I can easily generate a new ID each time the user adds a node.

```
// example
string1
string2
string3
```

The problem occurred because I was calling the `uniqueId` function inside another function which was also setting a new state (i.e. adding a new node), so I ended up in a race condition.

Sorry for the trouble.
